stackslib/src/net/download/nakamoto.rs

```diff
@@ -0,0 +1,4283 @@
+        // true
+        let prev_wanted_rc = prev_wanted_tenures
+            .first()
+            .map(|wt| {
+                pox_constants
+                    .block_height_to_reward_cycle(first_burn_height, wt.burn_height)
+                    .expect("FATAL: wanted tenure before system start")
+            })
+            .unwrap_or(u64::MAX);
+
+        let cur_wanted_rc = prev_wanted_rc.saturating_add(1);
+
+        let mut has_prev_inv = false;
+        let mut has_cur_inv = false;
+        for inv in inventory_iter {
+            if prev_wanted_rc < first_nakamoto_rc {
+                // assume the epoch 2.x inventory has this
+                has_prev_inv = true;
+            } else if inv.tenures_inv.get(&prev_wanted_rc).is_some() {
+                has_prev_inv = true;
+            }
+
+            if cur_wanted_rc < first_nakamoto_rc {
+                // assume the epoch 2.x inventory has this
+                has_cur_inv = true;
+            } else if inv.tenures_inv.get(&cur_wanted_rc).is_some() {
+                has_cur_inv = true;
+            }
+        }
+
+        if !has_prev_inv || !has_cur_inv {
+            test_debug!("No peer has an inventory for either the previous ({},{}) or current ({},{}) wanted tenures", prev_wanted_rc, has_prev_inv, cur_wanted_rc, has_cur_inv);
+            return true;
+        }
+
+        // the state machine updates `tenure_block_ids` _after_ `wanted_tenures`, so verify that
+        // this isn't a stale `tenure_block_ids` by checking that it contains at least one block in
+        // the prev_wanted_rc and at least one in the cur_wanted_rc
+        let mut has_prev_rc_block = false;
+        let mut has_cur_rc_block = false;
+        for (_naddr, available) in tenure_block_ids.iter() {
+            for (_ch, tenure_info) in available.iter() {
+                if tenure_info.start_reward_cycle == prev_wanted_rc
+                    || tenure_info.end_reward_cycle == prev_wanted_rc
+                {
+                    has_prev_rc_block = true;
+                }
+                if tenure_info.start_reward_cycle == cur_wanted_rc
+                    || tenure_info.end_reward_cycle == cur_wanted_rc
+                {
+                    has_cur_rc_block = true;
+                }
+            }
+        }
+
+        if (prev_wanted_rc >= first_nakamoto_rc && !has_prev_rc_block)
+            || (cur_wanted_rc >= first_nakamoto_rc && !has_cur_rc_block)
+        {
+            test_debug!(
+                "tenure_block_ids stale: missing representation in reward cycles {} ({}) and {} ({})",
+                prev_wanted_rc,
+                has_prev_rc_block,
+                cur_wanted_rc,
+                has_cur_rc_block,
+            );
+            return true;
+        }
+
+        let mut ret = false;
+        for (_naddr, available) in tenure_block_ids.iter() {
+            for wt in prev_wanted_tenures.iter() {
```

Comment:
`wt` has its own `processed` field. What is that used for, if not used in `has_unprocessed_tenures`?

Reply:
`.processed` is set if the tenure's data is already present in the staging blocks DB, so it doesn't need to be downloaded in the first place.  `have_unprocessed_tenures()`  determines whether or not all _available_ tenures (a subset of the list of wanted tenures) from the node's peers have been downloaded, even if they haven't yet been stored.  The downloader state machine looks at wanted tenures with `.processed = false` to construct `TenureStartEnd`s using the peer's neighbors' inventory vectors, and calls `have_unprocessed_tenures()` to see whether or not all of these `TenureStartEnd`s' data have been fetched.